src/analyses/variable-sensitivity/abstract_environment.cpp

```diff
@@ -53,59 +54,88 @@ std::vector<abstract_object_pointert> eval_operands(
   const abstract_environmentt &env,
   const namespacet &ns);
 
+bool is_ptr_diff(const exprt &expr)
+{
+  return (expr.id() == ID_minus) &&
+         (expr.operands()[0].type().id() == ID_pointer) &&
+         (expr.operands()[1].type().id() == ID_pointer);
+}
+
+bool is_ptr_comparison(const exprt &expr)
+{
+  auto const &id = expr.id();
+  if(id == ID_not)
+  {
+    auto const &not_expr = to_not_expr(expr);
+    return is_ptr_comparison(not_expr.op());
+  }
+
+  bool is_comparison = id == ID_equal || id == ID_notequal || id == ID_lt ||
+                       id == ID_le || id == ID_gt || id == ID_ge;
+
+  return is_comparison && (expr.operands()[0].type().id() == ID_pointer) &&
+         (expr.operands()[1].type().id() == ID_pointer);
+}
+
+static bool is_access_expr(const irep_idt &id)
+{
+  return id == ID_member || id == ID_index || id == ID_dereference;
+}
+
+static bool is_object_creation(const irep_idt &id)
+{
+  return id == ID_array || id == ID_struct || id == ID_constant ||
+         id == ID_address_of;
+}
+
+static bool is_dynamic_allocation(const exprt &expr)
+{
+  return expr.id() == ID_side_effect && expr.get(ID_statement) == ID_allocate;
+}
+
 abstract_object_pointert
 abstract_environmentt::eval(const exprt &expr, const namespacet &ns) const
 {
   if(bottom)
     return abstract_object_factory(expr.type(), ns, false, true);
 
   // first try to canonicalise, including constant folding
-  const exprt &simplified_expr = simplify_expr(expr, ns);
+  const exprt &simplified_expr = simplify_vsd_expr(expr, ns);
```

Comment:
Pondering : do we have to simplify here?  Could we do special case recognition first?

Reply:
We have to customise the simplification, because it recurses right down the expression tree. If we just let the simplification run, then the special case ends up being eliminated before we get a chance to try and evaluate it.